File: ios/PlayKit/PlayKit/PlayKitInteractor.m
//
//  PlayKitInteractor.m
//  PlayKit
//
//  Created by Lamine Ndiaye on 21/06/2019.
//  Copyright © 2019 Kaltura. All rights reserved.
//

#import "PlayKitInteractor.h"
#import <AVFoundation/AVFoundation.h>

@interface PlayKitInteractor()
@property (strong, nonatomic) id<Player> player;
@end

@implementation PlayKitInteractor

+(instancetype)playKitInteractor:(id<Player>) player {
    PlayKitInteractor *interactor = [[PlayKitInteractor alloc] init];
    interactor.player = player;
    return interactor;
}

- (NSArray<NSValue *> * _Nonnull)loadedTimeRanges {
    if (self.player == nil) {
        return [NSArray array];
    }
    
    NSMutableArray *timeRanges = [NSMutableArray array];
    for (PKTimeRange *range in [self.player loadedTimeRanges]) {
        TimeRange *timeRange = [[TimeRange alloc] initWithStart: range.start duration: range.duration];
        [timeRanges addObject:[[NSValue alloc] initWithTimeRange:timeRange]];
    }
    
    return timeRanges;
}

- (double)playbackTime {
    return self.player.currentTime;
}

- (void)updatePeakBitRate:(double)bitRate {
    self.player.settings.network.preferredPeakBitRate = bitRate;
}

- (double)bufferTarget {
    return self.player.settings.network.preferredForwardBufferDuration;
}

- (void)setBufferTarget:(double)target {
    self.player.settings.network.preferredForwardBufferDuration = target;
}

@end

File: ios/PhoenixAnalyticsSample/PhoenixAnalyticsSample-Swift/PlayKitInteractor.swift
//
//  PlayKitInteractor.swift
//  PlayKitOVPStarter
//
//  Created by Jean-Baptiste Louazel on 13/06/2019.
//  Copyright © 2019 Kaltura. All rights reserved.
//

import PlayKit
import StreamrootSDK

// Interactor between Streamroot and Kaltura PlayKit
// Binds Streamroot's events with player API
class PlayKitInteractor: DNAClientDelegate {
    private let player: Player
    
    public init(_ player: Player) {
        self.player = player
    }

    func updatePeakBitRate(_ bitRate: Double) {
        player.settings.network.preferredPeakBitRate = bitRate
    }
    
    func playbackTime() -> Double {
        return player.currentTime
    }
    
    func loadedTimeRanges() -> [NSValue] {
        return (player.loadedTimeRanges ?? [])
            .map { NSValue(timeRange: TimeRange(start: $0.start, duration: $0.duration)) }
    }
    
    func setBufferTarget(_ target: Double) {
        player.settings.preferredForwardBufferDuration = target
    }
    
    func bufferTarget() -> Double {
        return player.settings.preferredForwardBufferDuration
    }
}
